samples/java/src/main/java/com/modulr/api/ModulrApiAuth.java

```diff
@@ -0,0 +1,114 @@
+package com.modulr.api;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SignatureException;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.*;
+
+public class ModulrApiAuth {
+    private static final String HMAC_SHA1_ALGORITHM = "HmacSHA1";
+    private static final String DATE_PATTERN = "EEE, dd MMM yyyy HH:mm:ss z";
+    private final String secret;
+    private final String token;
+    private Date date;
+
+    private String lastUsedNonce;
+
+    public ModulrApiAuth(String token, String secret) {
+        this.token = token.trim();
+        this.secret = secret.trim();
+    }
+
+    public Map<String, String> generateApiAuthHeaders(String nonce) {
+        return buildHeaders(nonce, false);
+    }
+
+    public Map<String, String> generateRetryApiAuthHeaders() {
+        return buildHeaders(this.lastUsedNonce, true);
+    }
+
+    private Map<String, String> buildHeaders(String nonce, Boolean retry) {
+        final Map<String, String> headerParams = new HashMap<>();
+        try {
+            String hmac = generateHmac(nonce);
+
+            headerParams.put("Authorization", formatAuthHeader(this.token, hmac));
+            headerParams.put("Date", getFormattedDate(this.getDate()));
+            headerParams.put("x-mod-nonce", nonce);
+            headerParams.put("x-mod-retry", String.valueOf(retry));
+
+            this.lastUsedNonce = nonce;
+        } catch (SignatureException e) {
+            e.printStackTrace();
```

Comment:
probably best to throw IllegalStateException so that the client doesnt get silent failures

Reply:
Fixed it now.